src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -2399,10 +2422,19 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                     }
             }
 
-            if (traceOpcodes) EndInstructionTrace(gasAvailable, vmState.Memory?.Size ?? 0);
+            if (traceOpcodes)
+            {
+                UpdateCurrentState(vmState, programCounter, gasAvailable, stack.Head);
+                EndInstructionTrace(gasAvailable, vmState.Memory?.Size ?? 0);
+            }
         }
 
         UpdateCurrentState(vmState, programCounter, gasAvailable, stack.Head);
+
+        if (traceOpcodes && debugMode.IsOn)
+        {
+            debugMode.Debugger.TryWait(vmState);
```

Comment:
why there need to be 2x TryWait calls in a loop?

Reply:
the 2nd call to ``Debugger.TryWait`` happens after the loop ends.

Reply:
so we can capture the state after running the last opcode (the final state)
![image](https://user-images.githubusercontent.com/59487204/236118976-0c22fa07-f85e-411f-9f30-622e5ce80dcd.png)

Reply:
Wouldn't suffice then call after the loop?